metals/src/main/scala/scala/meta/internal/rename/RenameProvider.scala

```diff
@@ -100,9 +102,48 @@ final class RenameProvider(
   ): Future[WorkspaceEdit] = {
     val source = params.getTextDocument.getUri.toAbsolutePath
     compilations.compilationFinished(source).flatMap { _ =>
-      definitionProvider
+      val defininionFuture = definitionProvider
         .definition(source, params, token)
-        .flatMap { definition =>
+      defininionFuture.flatMap { definition =>
+        if (definition.symbol.isLocal) {
+
+          // We have to check if local symbol doesn't override any non-local symbols
+          // If it does, we can't use PC for local rename
+          val overridenSymbols =
+            (for {
+              semantic <- definition.semanticdb
+              symbolInfo <- semantic.symbols
+                .find(_.symbol == definition.symbol)
+              overridenSymbols = symbolInfo.overriddenSymbols
+              if symbolInfo.overriddenSymbols.nonEmpty
+              occ = semantic.occurrences
+                .filter(occ => overridenSymbols.contains(occ.symbol))
+            } yield occ).getOrElse(Nil)
+
+          overridenSymbols
+            .find(!_.symbol.isLocal)
+            .map(oSym =>
+              this.rename(
+                new RenameParams(
+                  params.getTextDocument,
+                  oSym.getRange.toLsp.getStart(),
+                  params.getNewName(),
+                ),
+                token,
+              )
+            )
+            .getOrElse(
+              compilers
+                .rename(params, token)
+                .map(_.asScala.toList)
+                .map { edits =>
+                  new WorkspaceEdit(
+                    documentEdits(Map(source -> edits)).asJava
+                  )
+                }
+            )
+
+        } else {
```

Comment:
Might be a bit simpler check that together in the first if.
```suggestion
      defininionFuture.flatMap { definition =>
        // We have to check if local symbol doesn't override any non-local symbols
        // If it does, we can't use PC for local rename
        def overridesNonLocal =  (for {
              semantic <- definition.semanticdb
              symbolInfo <- semantic.symbols
                .find(_.symbol == definition.symbol)
              overridenSymbols = symbolInfo.overriddenSymbols
            } yield overridenSymbols.exists(!_.isLocal)).getOrElse(true)
        if (definition.symbol.isLocal && !overridesNonLocal) {
              compilers
                .rename(params, token)
                .map(_.asScala.toList)
                .map { edits =>
                  new WorkspaceEdit(
                    documentEdits(Map(source -> edits)).asJava
                  )
                }
            )
        } else {
```
Also this shouldn't be needed:
```
semantic.occurrences
                .filter(occ => overridenSymbols.contains(occ.symbol))
```

Reply:
It looks way better, thanks.  I thought we have to run rename on nonlocal overriding symbol to make it work, but apparently it's not needed